### ISSUE TITLE
Lua 5.4 features suggestions

### ISSUE THREAD

**Issue by v1993** [pavouk/lgi#230]
As Lua 5.4 release is slowly getting closer, there's a list of features that would be nice to have in it:

* #94, but with new warnings system.
* `<toclose>` support for major things like objects. It should act like garbage collection, cleaning object. Any further access to object should result in an error.

While first feature isn't that important, second one would be very useful to have.

**Comment by v1993** [comment]
Interesting. If "make objects unusable in __gc metamethod" in last release is true, then it should be possible to simply use it in `__close` metamethod.

**Comment by psychon** [comment]
No. If you do `getmetatable(obj).__gc(obj)`, then the C object is released, but its address is still remembered in Lua and tied to the (now broken) Lua wrapper userdata. If later another object is allocated at the same address, then LGI will just return the "already existing" wrapper userdata instead of collecting a new one.

(This issue just bit me in some piece of code where I was trying to be clever)

The same problem would occur when the `__gc` metamethod is used as `__close`. LGI needs to be changed to "clean up" the table with the mappers to make this safe.

This cache can be found in `lgi_object_2lua` and would need to be cleaned up in `object_gc` (but only if it still exists - I am not sure what would happen during `lua_close()`...). And of course, such a change would be needed not only for objects, but also for other things that are cached. `git grep lgi_cache_create` says that callables and records might be affected as well.
